### PR TITLE
Begin porting from Python2/Django 1.11 to Python3/Django 2.0

### PR DESCRIPTION
**Pull request by speizerj**
Changes the Dockerfile to use Python 3.6, and requirements.txt to use Django 2.0

Clean up exceptions to Python3 format in dashboard/helpers.py

### CODE REVIEW

**Comment by owocki** [comment]
awesome! a few questions:

1. mind resolving the conflicts?  
1. whats left TODO here?  
1. should i ask the community to test this yet?

**Comment by speizerj** [comment]
> mind resolving the conflicts?
yep, will do

> whats left TODO here?
I need to get the app to a point where it can run the local server successfully before we can test. I'll have a better time estimate after working on it some more today.

> should i ask the community to test this yet?
hold off for now, once the app is running then I'll need the help

**Comment by speizerj** [comment]
@owocki the local server is now running successfully on python3/django2. I've tested out the lifecycle of a funded issue on Ropsten and it's working for me (funding, claiming, accepting, rejecting)

Tips appear to be broken and so that's what I'll work on next, but if anyone has any insight there I'm all ears. The request.body is empty in send_tip_2 view, not sure why.

Now would be a good time to open it up to testing - I'm sure there are plenty of edge cases.

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM python:2.7-slim
+FROM python:3.6
```
Comment (mbeacom) [inline Dockerfile]: Can you pin this to `3.6-slim-stretch`

Comment (mbeacom) [inline Dockerfile]: Would it be possible to switch this to `3.6-slim-jessie` ?

**Comment by owocki** [comment]
just posted about this to #focus-dev-python on slack.

also, looks like theres a bunch new merge conflicts. might be because of the isort PR.  sorry!

**Comment by mbeacom** [comment]
My apologies! Most, if not all of those conflicts should be fine to resolve if you drop in the `setup.cfg` and run `sort -rc --atomic .` in the repo root: `web` @speizerj

**Comment by owocki** [comment]
installing the dependancies so i can test now..

**Comment by owocki** [comment]
installing the dependancies so i can test now..

**Comment by owocki** [comment]
got it running; i similarly get the tips error

**Comment by mbeacom** [comment]
The failure on `send_tip_2` is because `b''` is being passed as `request.body` and falsely passing through `if request.body != '':` since it's true.
Not saying this is the right answer, but `if request.body not in ['', b'']:` at least eliminates the `JSONDecodeError`

File: Dockerfile (hunk 1 above)
Comment (speizerj) [reply Dockerfile]: sure thing

**Comment by speizerj** [comment]
dockerfile has been updated and merge conflicts resolved manually - working on the tip bug now

**Comment by speizerj** [comment]
Tips appear to be working now, though I could only test out the sending functionality. The receive page appears to work as well but I couldn't confirm because I don't have the proper setup to test out the email functionality, and I couldn't get a link to test.

File: Dockerfile [hunk 2]
```diff
@@ -1,12 +1,12 @@
-FROM python:2.7-slim
+FROM python:3.6-slim-jessie
```
Comment (jasonrhaas) [inline Dockerfile]: Any reason you chose the `-slim-jessie` image over the `-slim` or just the regular `python:3.6` image?  I'm wondering if some of the linux packages you are installing below might already be included in one of the other python images.  Might speed up the build time a bit.

File: app/dashboard/helpers.py [hunk 3]
```diff
@@ -67,7 +67,7 @@ def title(request):
     urlVal = URLValidator()
     try:
         urlVal(url)
-    except ValidationError, e:
+    except ValidationError as e:
```
Comment (jasonrhaas) [inline app/dashboard/helpers.py]: Total nit pick but if you aren't using any information from the Error Message you shouldn't even need the `as e` part.

File: app/dashboard/helpers.py [hunk 4]
```diff
@@ -77,7 +77,7 @@ def title(request):
 
     try:
         html_response = requests.get(url)
-    except ValidationError, e:
+    except ValidationError as e:
```
Comment (jasonrhaas) [inline app/dashboard/helpers.py]: Same comment as above.

File: app/dashboard/helpers.py [hunk 5]
```diff
@@ -96,7 +96,7 @@ def title(request):
             for link in soup.find_all('h1'):
                 print(link.text)
 
-    except ValidationError, e:
+    except ValidationError as e:
```
Comment (jasonrhaas) [inline app/dashboard/helpers.py]: Same comment as above.

File: app/dashboard/helpers.py [hunk 6]
```diff
@@ -119,7 +119,7 @@ def keywords(request):
     urlVal = URLValidator()
     try:
         urlVal(url)
-    except ValidationError, e:
+    except ValidationError as e:
```
Comment (jasonrhaas) [inline app/dashboard/helpers.py]: Same comment as above.

File: app/dashboard/helpers.py [hunk 7]
```diff
@@ -138,7 +138,7 @@ def keywords(request):
         keywords.append(split_repo_url[-2])
 
         html_response = requests.get(repo_url)
-    except ValidationError, e:
+    except ValidationError as e:
```
Comment (jasonrhaas) [inline app/dashboard/helpers.py]: Same comment as above.

File: app/dashboard/helpers.py [hunk 8]
```diff
@@ -149,7 +149,7 @@ def keywords(request):
         for ele in eles:
             keywords.append(ele.text)
 
-    except ValidationError, e:
+    except ValidationError as e:
```
Comment (jasonrhaas) [inline app/dashboard/helpers.py]: Same comment as above.

File: app/dashboard/views.py [hunk 9]
```diff
@@ -53,7 +53,7 @@ def send_tip(request):
 @ratelimit(key='ip', rate='2/m', method=ratelimit.UNSAFE, block=True)
 def receive_tip(request):
 
-    if request.body != '':
+    if request.body:
```
Comment (jasonrhaas) [inline app/dashboard/views.py]: 👍

**Comment by owocki** [comment]
thanks for pushing this fwd yall.  heads down on other stuff today, but hit me up on slack if you want more community involvement

File: app/dashboard/helpers.py (hunk 3 above)
Comment (mbeacom) [reply app/dashboard/helpers.py]: 💯

File: Dockerfile (hunk 2 above)
Comment (mbeacom) [reply Dockerfile]: Using the slim base cuts this docker image size in half.

Comment (jasonrhaas) [reply Dockerfile]: Right -- I usually use the `slim` versions also unless I have some issue installing things on the container.  Makes sense then re: openssl version.

**Comment by speizerj** [comment]
@owocki I think it's ready for community help with testing. My experience with the app is limited so I tested what I could (and everything was working), but some help would be awesome.

**Comment by owocki** [comment]
posted in the python channel.  i hope to find some time to test myself tomorrow

**Comment by mbeacom** [comment]
Via http://localhost:8000/whitepaper submitting request form:
```Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 128, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.6/site-packages/ratelimit/decorators.py", line 30, in _wrapped
    return fn(*args, **kw)
  File "/code/app/tdi/views.py", line 76, in whitepaper_new
    send_mail(settings.CONTACT_EMAIL, settings.CONTACT_EMAIL, "New Whitepaper Request", str(body))
  File "/code/app/marketing/mails.py", line 30, in send_mail
    get_or_save_email_subscriber(to_email, 'internal')
  File "/code/app/marketing/utils.py", line 48, in get_or_save_email_subscriber
    es.set_priv()
  File "/code/app/marketing/models.py", line 47, in set_priv
    h.update("{}-{}-{}".format(h.hexdigest(), timezone.now(), self.email))```

File: app/app/urls.py [hunk 10]
```diff
@@ -28,7 +28,8 @@
 import retail.views
 import tdi.views
 from dashboard.router import router
-from sitemap import sitemaps
+from django.contrib.sitemaps.views import sitemap
+from . import sitemaps
```
Comment (mbeacom) [inline app/app/urls.py]: This breaks the sitemaps.  This should probably be `from .sitemaps import sitemaps`

**Comment by owocki** [comment]
> Via http://localhost:8000/whitepaper submitting request form:

Whats the type of the exception on this one?  if its slack API key or DB migration related, its possible this exception is not django related.

File: app/app/urls.py (hunk 10 above)
Comment (owocki) [reply app/app/urls.py]: 👍

**Comment by mbeacom** [comment]
That could be. I'll add some valid entries to my local instance and see if that resolves the issue.  If that's the case, it **should** simply be the `sitemap` change.

**Comment by owocki** [comment]
i should have some time to help test/ deploy this on stage later this week. let me know when its ready

File: app/marketing/models.py [hunk 11]
```diff
@@ -44,7 +44,7 @@ def set_priv(self):
         from django.utils import timezone
 
         h = hashlib.new('ripemd160')
-        h.update("{}-{}-{}".format(h.hexdigest(), timezone.now(), self.email))
+        h.update(b"{}-{}-{}".format(h.hexdigest(), timezone.now(), self.email))
```
Comment (speizerj) [inline app/marketing/models.py]: @mbeacom this fixes the 500 error you were getting - I believe the python3 hashlib requires a bytestring now

**Comment by speizerj** [comment]
The above error on http://localhost:8000/whitepaper is resolved but I don't have sendgrid credentials so I'm obviously erroring at that point. I may have an old sendgrid account lying around I can use, but if not, can someone else confirm it's now working?

**Comment by speizerj** [comment]
FYI here's the error I'm getting when trying to submit the /whitepaper form (looks to be sendgrid related - assuming it's lack of credentials):

```
Traceback:

File "/usr/local/lib/python3.6/site-packages/python_http_client/client.py" in _make_request
  157.             return opener.open(request)

File "/usr/local/lib/python3.6/urllib/request.py" in open
  532.             response = meth(req, response)

File "/usr/local/lib/python3.6/urllib/request.py" in http_response
  642.                 'http', request, response, code, msg, hdrs)

File "/usr/local/lib/python3.6/urllib/request.py" in error
  570.             return self._call_chain(*args)

File "/usr/local/lib/python3.6/urllib/request.py" in _call_chain
  504.             result = func(*args)

File "/usr/local/lib/python3.6/urllib/request.py" in http_error_default
  650.         raise HTTPError(req.full_url, code, msg, hdrs, fp)

During handling of the above exception (HTTP Error 401: Unauthorized), another exception occurred:

File "/usr/local/lib/python3.6/site-packages/django/core/handlers/exception.py" in inner
  35.             response = get_response(request)

File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  128.                 response = self.process_exception_by_middleware(e, request)

File "/usr/local/lib/python3.6/site-packages/django/core/handlers/base.py" in _get_response
  126.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/usr/local/lib/python3.6/site-packages/ratelimit/decorators.py" in _wrapped
  30.             return fn(*args, **kw)

File "/code/app/tdi/views.py" in whitepaper_new
  76.     send_mail(settings.CONTACT_EMAIL, settings.CONTACT_EMAIL, "New Whitepaper Request", str(body))

File "/code/app/marketing/mails.py" in send_mail
  62.     response = sg.client.mail.send.post(request_body=mail.get())

File "/usr/local/lib/python3.6/site-packages/python_http_client/client.py" in http_request
  227.                 return Response(self._make_request(opener, request))

File "/usr/local/lib/python3.6/site-packages/python_http_client/client.py" in _make_request
  161.             raise exc

Exception Type: UnauthorizedError at /whitepaper
Exception Value: HTTP Error 401: Unauthorized
```

**Comment by owocki** [comment]
you can generate a free sendgrid key at sendgrid.com (or just comment out that line for testing)

**Comment by speizerj** [comment]
Whitepaper generation is now working for me locally, though I had to test with a dummy PDF (I don't have the whitepaper source). The watermark is being generated but does appear to be cut off a bit on the right hand side - not sure if that's a new bug or existing.

**Comment by owocki** [comment]
looks like theres a few merge conflicts on this PR now.   we should go ahead and get this merged so that the overhead of keeping it current is less :P

**Comment by speizerj** [comment]
no worries @owocki I don't mind keeping it current - I'll take care of those merge conflicts tonight.

Sounds good re: QA/merge - let me know how I can help

**Comment by mbeacom** [comment]
I'm going to be raising a PR for GH integration and bounty model changes this evening, so I should be able to run through this again at some point shortly.  Anything specific that I should hit or just run down the views and hit some of the common methods? I was hoping to get in some general tests to help you out, but didn't get that far yet! Maybe after the GH/bounty changes!

**Comment by owocki** [comment]
derp.  looks like theres some merge conflcits now..

i could likely carve out some time to nail this down on weds if anyone else wants to do that with me

**Comment by speizerj** [comment]
sure thing - I'll take care of the conflicts tonight and then happy to join you on Wed

**Comment by mbeacom** [comment]
We should be good with GH/express interest before then, so I should be free to help you both test/resolve whatever is needed as well (and you'll probably have more merge conflicts :trollface: )

Let me know if either of you need any help with this before then!

**Comment by jasonrhaas** [comment]
hows it going on the move from Python 2 --> 3?

**Comment by owocki** [comment]
its... fallen to the backburner unforgtunately... :(

**Comment by owocki** [comment]
closing this as it's been moved over to the `january2018_feature_integration` branch.  @mbeacom @speizerj are there any ongoing topics of converstaion from this branch you want to carry over there?

**Comment by speizerj** [comment]
Nope I think we should be good, the only outstanding items were keeping it current with regards to merge conflicts and testing out all the flows to ensure there were no issues. I'll follow up on the new PR

**Comment by owocki** [comment]
whoops.. meant to post this to another issue.. cleaning this up now

**Pull request closed without merging**